core/src/runtime/mod.rs

```diff
@@ -84,6 +86,10 @@ pub struct Runtime {
     pub max_syscall_cycles: u32,
 
     pub emit_events: bool,
+
+    pub pi_hasher: Option<Sha256>,
```

Comment:
nit: seems unecessary, can you just instantiate a Sh256 instance locally where you need it?

Reply:
I don't keep a buffer for the public input bytes.  Instead, it's streamed into this hasher.  lmk if you think the raw bytes should be buffered instead.

Reply:
I do feel like it would be better if raw bytes were buffered instead, that way we can inspect the bytes more easily and avoid doing this take/unwrap thing